File: src/main/java/com/uiptv/db/SQLConnection.java
package com.uiptv.db;

import com.uiptv.util.ConfigFileReader;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;

import static com.uiptv.util.Platform.getUserHomeDirPath;
import static com.uiptv.util.StringUtils.isNotBlank;

public class SQLConnection {
    private static final int BUSY_TIMEOUT_MS = 10_000;
    private static final int CACHE_SIZE_KIB = 32_000;
    private static final int INIT_RETRY_ATTEMPTS = 6;
    private static final long JOURNAL_SIZE_LIMIT_BYTES = 16L * 1024L * 1024L;
    private static final double VACUUM_FREE_PAGE_THRESHOLD = 0.15d;
    private static final int WAL_AUTOCHECKPOINT_PAGES = 1_000;
    private static final long INIT_RETRY_DELAY_MS = 250L;

    private static String databasePathFromConfigFile = ConfigFileReader.getDbPathFromConfigFile();
    private static String dbPath = isNotBlank(databasePathFromConfigFile) ? databasePathFromConfigFile : getUserHomeDirPath() + File.separator + "uiptv.db";

    private SQLConnection() {
    }

    static {
        init();
    }

    @SuppressWarnings("java:S1141")
    public static synchronized void init() {
        try {
            FileUtils.touch(new File(dbPath));
            for (int attempt = 1; attempt <= INIT_RETRY_ATTEMPTS; attempt++) {
                try (Connection conn = openConnection()) {
                    applyPersistentPragmas(conn);
                    applySchema(conn);
                    runStartupMaintenance(conn);
                    return;
                } catch (SQLException ex) {
                    if (isBusy(ex) && attempt < INIT_RETRY_ATTEMPTS) {
                        sleepBeforeRetry();
                        continue;
                    }
                    throw new DatabaseAccessException("Unable to initialize database migrations", ex);
                }
            }
        } catch (IOException e) {
            throw new DatabaseAccessException("Unable to create database file", e);
        }
    }

    private static void applySchema(Connection conn) throws SQLException {
        if (DatabasePatchesUtils.hasMigrationsListResource()) {
            DatabasePatchesUtils.applyPatches(conn);
            return;
        }
        if (DatabasePatchesUtils.hasBaselineResource()) {
            DatabasePatchesUtils.applyBaseline(conn);
            return;
        }
        throw new IllegalStateException("No database schema resources found (missing migrations list and baseline)");
    }

    public static synchronized void setDatabasePath(String path) {
        dbPath = path;
        init();
    }

    public static Connection connect() {
        try {
            return openConnection();
        } catch (SQLException e) {
            throw new DatabaseAccessException("Unable to open database connection", e);
        }
    }

    private static Connection openConnection() throws SQLException {
        Connection conn = DriverManager.getConnection("jdbc:sqlite:" + dbPath);
        try (Statement statement = conn.createStatement()) {
            statement.execute("PRAGMA busy_timeout = " + BUSY_TIMEOUT_MS);
            statement.execute("PRAGMA journal_mode = WAL");
            statement.execute("PRAGMA synchronous = NORMAL");
            statement.execute("PRAGMA temp_store = MEMORY");
            statement.execute("PRAGMA cache_size = -" + CACHE_SIZE_KIB);
            statement.execute("PRAGMA journal_size_limit = " + JOURNAL_SIZE_LIMIT_BYTES);
            statement.execute("PRAGMA wal_autocheckpoint = " + WAL_AUTOCHECKPOINT_PAGES);
        }
        return conn;
    }

    private static void applyPersistentPragmas(Connection conn) throws SQLException {
        try (Statement statement = conn.createStatement()) {
            statement.execute("PRAGMA auto_vacuum = INCREMENTAL");
        }
    }

    private static void runStartupMaintenance(Connection conn) throws SQLException {
        try (Statement statement = conn.createStatement()) {
            statement.execute("PRAGMA optimize");
            statement.execute("PRAGMA wal_checkpoint(PASSIVE)");
            if (shouldVacuum(statement)) {
                statement.execute("VACUUM");
                statement.execute("PRAGMA wal_checkpoint(TRUNCATE)");
            }
        }
    }

    private static boolean shouldVacuum(Statement statement) throws SQLException {
        long freePages = queryLongPragma(statement, "freelist_count");
        long totalPages = queryLongPragma(statement, "page_count");
        if (totalPages <= 0) {
            return false;
        }
        return ((double) freePages / totalPages) > VACUUM_FREE_PAGE_THRESHOLD;
    }

    private static long queryLongPragma(Statement statement, String pragmaName) throws SQLException {
        try (var resultSet = statement.executeQuery("PRAGMA " + pragmaName)) {
            return resultSet.next() ? resultSet.getLong(1) : 0L;
        }
    }

    private static boolean isBusy(SQLException exception) {
        SQLException current = exception;
        while (current != null) {
            String message = current.getMessage();
            if (current.getErrorCode() == 5 || (message != null && message.contains("SQLITE_BUSY"))) {
                return true;
            }
            current = current.getNextException();
        }
        return false;
    }

    private static void sleepBeforeRetry() {
        try {
            Thread.sleep(INIT_RETRY_DELAY_MS);
        } catch (InterruptedException interruptedException) {
            Thread.currentThread().interrupt();
            throw new DatabaseAccessException("Interrupted while waiting to retry database initialization", interruptedException);
        }
    }

}

File: server/src/main/kotlin/com/uiptv/db/SQLConnection.kt
package com.uiptv.db

import com.zaxxer.hikari.HikariDataSource
import org.jetbrains.exposed.sql.Database
import java.io.IOException
import java.sql.Connection
import java.sql.SQLException

object SqlConnectionRuntime {
    private const val BUSY_TIMEOUT_MS = 10_000
    private const val INIT_RETRY_ATTEMPTS = 6
    private const val INIT_RETRY_DELAY_MS = 250L

    @Volatile
    private var exposedDatabase: Database? = null

    @Volatile
    private var hikariDataSource: HikariDataSource? = null

    init {
        init()
    }

    @JvmStatic
    @Synchronized
    fun init() {
        try {
            ensureDatabasePathReady()
            rebuildDataSource()
            for (attempt in 1..INIT_RETRY_ATTEMPTS) {
                try {
                    migrate()
                    exposedDatabase = Database.connect(dataSource())
                    return
                } catch (ex: SQLException) {
                    if (DatabaseRetrySupport.isBusy(ex) && attempt < INIT_RETRY_ATTEMPTS) {
                        DatabaseRetrySupport.sleepBeforeRetry(INIT_RETRY_DELAY_MS)
                        continue
                    }
                    throw DatabaseAccessException("Unable to initialize database migrations", ex)
                }
            }
        } catch (e: IOException) {
            throw DatabaseAccessException("Unable to create database file", e)
        }
    }

    @JvmStatic
    @Synchronized
    fun setDatabasePath(path: String) {
        if (path == databasePath()) {
            return
        }
        close()
        DatabasePathState.override(path)
        init()
    }

    @JvmStatic
    fun databasePath(): String = DatabasePathState.currentPath()

    @JvmStatic
    fun connect(): Connection =
        try {
            dataSource().connection.also { DatabaseSqliteSupport.configurePragmas(it, BUSY_TIMEOUT_MS) }
        } catch (e: SQLException) {
            throw DatabaseAccessException("Unable to open database connection", e)
        }

    @JvmStatic
    fun database(): Database = exposedDatabase ?: synchronized(this) {
        exposedDatabase ?: Database.connect(dataSource()).also { exposedDatabase = it }
    }

    @JvmStatic
    @Synchronized
    fun close() {
        exposedDatabase = null
        hikariDataSource?.close()
        hikariDataSource = null
    }

    @JvmStatic
    @Synchronized
    fun dataSource(): HikariDataSource {
        val existing = hikariDataSource
        if (existing != null && !existing.isClosed) {
            return existing
        }
        ensureDatabasePathReadyUnchecked()
        rebuildDataSource()
        return hikariDataSource ?: throw IllegalStateException("Data source not initialized")
    }

    private fun migrate() {
        DatabaseMigrationRunner.migrate(
            dbPath = databasePath(),
            openConnection = ::openConnection
        )
    }

    @Throws(SQLException::class)
    private fun openConnection(): Connection {
        return DatabaseSqliteSupport.openConnection(databasePath(), BUSY_TIMEOUT_MS)
    }

    @Synchronized
    private fun rebuildDataSource() {
        ensureDatabasePathReadyUnchecked()
        hikariDataSource?.close()
        hikariDataSource = DatabasePoolFactory.create(
            databasePath(),
            DatabaseSqliteSupport.sqliteProperties(BUSY_TIMEOUT_MS)
        )
    }

    @Throws(IOException::class)
    private fun ensureDatabasePathReady() {
        DatabaseFileSupport.ensureDatabasePathReady(databasePath())
    }

    private fun ensureDatabasePathReadyUnchecked() {
        try {
            ensureDatabasePathReady()
        } catch (e: IOException) {
            throw DatabaseAccessException("Unable to create database file", e)
        }
    }

}
